python2/typing.py

```diff
@@ -1092,6 +1086,15 @@ def __new__(cls, name, bases, namespace,
                          ", ".join(str(g) for g in gvars)))
                 tvars = gvars
 
+        if extra is None:
+            extra = namespace.get('__extra__')
+        if extra is not None and type(extra) is abc.ABCMeta and extra not in bases:
+            bases = (extra,) + bases
+        bases = tuple(_gorg(b) if isinstance(b, GenericMeta) else b for b in bases)
+        if any(isinstance(b, GenericMeta) and b is not Generic for b in bases):
```

Comment:
This could use a comment explaining what's happening -- it looks like the idea is to remove bare Generic from the bases if there are other generic-derived bases, but to leave generic if if not.